src/hb-set.hh

```diff
@@ -366,14 +383,48 @@ struct hb_set_t
     dirty ();
     page->del (g);
   }
+
   void del_range (hb_codepoint_t a, hb_codepoint_t b)
   {
     /* TODO perform op even if !successful. */
-    /* TODO Optimize, like add_range(). */
     if (unlikely (!successful)) return;
-    for (unsigned int i = a; i < b + 1; i++)
-      del (i);
+    if (unlikely (a > b || a == INVALID || b == INVALID)) return;
+    dirty ();
+    unsigned int ma = get_major (a);
+    unsigned int mb = get_major (b);
+    /* Delete pages from ds through de if ds <= de. */
+    int ds = (a == major_start (ma))? (int)ma: (int)(ma + 1);
+    int de = (b + 1 == major_start (mb + 1))? (int)mb: ((int)mb - 1);
+    if (ds > de || (int)ma < ds)
+    {
+      page_t *page = page_for (a);
+      if (page)
+      {
+	if (ma == mb)
+	  page->del_range (a, b);
+	else
+	  page->del_range (a, major_start (ma + 1) - 1);
+      }
+    }
+    if (de < (int)mb && ma != mb)
+    {
+      page_t *page = page_for (b);
+      if (page)
+	page->del_range (major_start (mb), b);
+    }
+    if (ds <= de)
+    {
```

Comment:
The contents of this if could be extracted into a function, del_pages(ds, de) which would help improve readability of this method.